driver/src/main/java/org/neo4j/driver/internal/InternalStatementResult.java

```diff
@@ -274,46 +248,34 @@ else if ( done )
     @Override
     public <T> List<T> list( Function<Record, T> mapFunction )
     {
-        if ( isEmpty() )
-        {
-            assertOpen();
-            return emptyList();
-        }
-        else if ( position == -1 && hasNext() )
+        if ( hasNext() )
         {
             List<T> result = new ArrayList<>();
+
             do
             {
                 result.add( mapFunction.apply( next() ) );
             }
             while ( hasNext() );
 
-            consume();
             return result;
         }
         else
         {
-            throw new ClientException(
-                    format( "Can't retain records when cursor is not pointing at the first record (currently at position %d)", position )
-            );
+            return emptyList();
         }
     }
 
-    @SuppressWarnings("StatementWithEmptyBody")
     @Override
     public ResultSummary consume()
     {
-        if(!open)
-        {
-            return summary;
-        }
-
         while ( !done )
         {
             connection.receiveOne();
+            recordBuffer.clear();
```

Comment:
I think we have two issues here:

- We don't want to fill up the buffer during consumption as that would cause unnecessary memory pressure, hence we need to empty the buffer in the inner loop
- We also want to make sure we empty the buffer in any case - imagine consume being called while there are still records in the buffer but done is already true - hence it gets emptied one more time after the loop

Is there a better way?